urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/BasicOperations.java

```diff
@@ -0,0 +1,699 @@
+ * 
+ * Copyright (c) 2001-2011 Anders Moeller
+ * All rights reserved.
+ * 
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package dk.brics.automaton;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Basic automata operations.
+ */
+final public class BasicOperations {
+	
+	private BasicOperations() {}
+
+	/** 
+	 * Returns an automaton that accepts the concatenation of the languages of 
+	 * the given automata. 
+	 * <p>
+	 * Complexity: linear in number of states. 
+	 */
+	static public Automaton concatenate(Automaton a1, Automaton a2) {
+		if (a1.isSingleton() && a2.isSingleton())
+			return BasicAutomata.makeString(a1.singleton + a2.singleton);
+		if (isEmpty(a1) || isEmpty(a2))
+			return BasicAutomata.makeEmpty();
+		boolean deterministic = a1.isSingleton() && a2.isDeterministic();
+		if (a1 == a2) {
+			a1 = a1.cloneExpanded();
+			a2 = a2.cloneExpanded();
+		} else {
+			a1 = a1.cloneExpandedIfRequired();
+			a2 = a2.cloneExpandedIfRequired();
+		}
+		for (State s : a1.getAcceptStates()) {
+			s.accept = false;
+			s.addEpsilon(a2.initial);
+		}
+		a1.deterministic = deterministic;
+		a1.clearHashCode();
+		a1.checkMinimizeAlways();
+		return a1;
+	}
+	
+	/**
+	 * Returns an automaton that accepts the concatenation of the languages of
+	 * the given automata.
+	 * <p>
+	 * Complexity: linear in total number of states.
+	 */
+	static public Automaton concatenate(List<Automaton> l) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.